tests/unit_tests/scheduler/test_scheduler.py

```diff
@@ -204,6 +204,39 @@ async def wait():
     assert timeouteventfound
 
 
+@pytest.mark.parametrize("max_running", [0, 1, 2, 10])
+async def test_max_running(max_running, mock_driver, storage, tmp_path):
+    currently_running = 0
+    max_running_observed = 0
+
+    async def wait():
+        nonlocal currently_running
+        currently_running += 1
+        await asyncio.sleep(0.01)
+        currently_running -= 1
```

Comment:
We could change this so that instead of incrementing and decrementing `currently_running`, we can use a boolean list where `True` means started and `False` means stopped. We can then find `max_running_observed` by:
```py
async def wait():
    nonlocal runs
    runs.append(True)
    await asyncio.sleep(0.1)
    runs.append(False)

# ...

max_running_observed = 0
currently_running = 0
for run in runs:
    currently_running += 1 if run else -1
    max_running_observed = max(max_running_observed, currently_running)
```

Very advent of code. Also this way we don't rely on timings (other than `wait`'s `asyncio.sleep(0.1)`)

Reply:
Done!